### PR TITLE
Upgrade TensorFlow Probability with newer versions

### PR DESCRIPTION
**Pull request by jonas-eschle**
Needs https://github.com/spack/spack/pull/41069 (TF 0.23) for the newest version

### CODE REVIEW

**Comment by jonas-eschle** [comment]
@spackbot fix style

**Comment by jonas-eschle** [comment]
waiting for #41069

**Comment by jonas-eschle** [comment]
Ready from my side; I'll upgrade the versions once the deps are merged, could take a while. The TFP is so outdated however that an upgrade is necessary

**Comment by adamjstewart** [comment]
#41069 is very close now, just need to figure out why the CUDA build is failing. I opened an issue with the TF developers but no luck yet. If you know any Bazel maybe take a look at the error message and see if you have any ideas.

**Comment by jonas-eschle** [comment]
Thanks, glad to hear! I had my experience from the conda builds with TF build fixes (not as in "I know bazel", rather the opposite) but as it can take a while; let's see! (also it's not just about the TFP itself, but TFP is a dep of zfit in https://github.com/spack/spack/pull/42667 that I would actually like to bring in as a new package)

But let's see, thanks for the upgrade!

**Comment by jonas-eschle** [comment]
@adamjstewart what do you think, could we maybe move forward with this one, so I'll can also move forward with other recipes? I'll update it right afterwards and make a new PR, but seeing  what they write in TF, it could also take another month or more

**Comment by adamjstewart** [comment]
At a workshop right now but I'll take a closer look at TF and at your PR this weekend.

File: var/spack/repos/builtin/packages/py-tensorflow-probability/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-tensorflow-probability/package.py]: I vote we deprecate this version. It's ancient and has a custom URL syntax that makes the package look ugly.

File: var/spack/repos/builtin/packages/py-tensorflow-probability/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-tensorflow-probability/package.py]: I believe these should be >=, not ==. At least, that's the way things are encoded in `__init__.py`

File: var/spack/repos/builtin/packages/py-tensorflow-probability/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-tensorflow-probability/package.py:61]: Where do you see JAX/jaxlib listed as dependencies? Where do the versions come from? I would make a new section with a comment for the file that lists the dep versions so we can easily keep track of when they change.

File: var/spack/repos/builtin/packages/py-tensorflow-probability/package.py
Comment (jonas-eschle) [reply var/spack/repos/builtin/packages/py-tensorflow-probability/package.py]: I would disagree as it could break users code: old versions are _always bad_, so nobody should install it, I agree. But there is no cost in having it here, is there?
If you prefer to remove it, please go ahead, but I think in general "yanking" packages should only be done if _absolutely_ needed.

File: var/spack/repos/builtin/packages/py-tensorflow-probability/package.py
Comment (jonas-eschle) [reply var/spack/repos/builtin/packages/py-tensorflow-probability/package.py:61]: It's as much of a dependency as TF itself: TFP can be run with JAX or with TF and doesn't need one of them to run the other, AFAIU. jax is the pure python package, jaxlib is the compiled backend.

The versions come from the release notes of TFP which say against which version they have been tested; I've been propagating this since a while to the conda repository (https://github.com/conda-forge/tensorflow-probability-feedstock/blob/main/recipe/meta.yaml) and pinning the deps there (as conda doesn't have extras, additional deps are included by default).

Therefore, it seems to me as we either include both, TF _and_ jax, or neither. No strong opinion (and as mentioned, I am rather new to spack and not fully familiar with all the controls the end user has, the general vibes), can be fine both ways (in conda-forge we have both deps, but that's a different philosophy and more towards "it should work for the end user correctly")

File: var/spack/repos/builtin/packages/py-tensorflow-probability/package.py
Comment (jonas-eschle) [reply var/spack/repos/builtin/packages/py-tensorflow-probability/package.py]: They do, true, but in the [changelog](https://github.com/tensorflow/probability/releases/tag/v0.22.1) they say against which version things are tested and stable. The problem with keeping only a lower limit is that this may break versions with future TF releases as painfully happenend multiple times in the past. Therefore, we adopted in conda-forge a strict policy towards TF (and a laxer one towards JAX [after some discussion](https://github.com/conda-forge/tensorflow-probability-feedstock/issues/45)).

If you have a strong opinion on it, please go ahead and adjust (or let me know), both have advantages and disadvantages, so fine with both, but if we choose to relax, the question is what to do if users run into it? Add upper boundaries? If so where? Should we test all versions of TFP against TF, something that not even TFP seems to be doing?

**Review by jonas-eschle — COMMENTED** [review]
Thanks for taking a look at this!

File: var/spack/repos/builtin/packages/py-tensorflow-probability/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-tensorflow-probability/package.py]: This is why we have a deprecation policy. We aren't deleting it immediately, just deprecating it for a full release cycle before removing. https://spack.readthedocs.io/en/latest/packaging_guide.html#deprecating-old-versions describes the policy in more detail.

I don't have strong opinions about this, and I don't mind keeping it if anyone still uses it.

File: var/spack/repos/builtin/packages/py-tensorflow-probability/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-tensorflow-probability/package.py]: I don't have strong opinions here. I've also needed to pin some things related to TF/JAX in other packages because they frequently break backwards compatibility. One of the main advantages of Spack over pip/conda is that it's trivially easy to add upper bounds to ancient versions if users discover incompatibilities. So some of the conda-forge discussion may not exactly apply here. I'll leave the decision up to you.

File: var/spack/repos/builtin/packages/py-tensorflow-probability/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-tensorflow-probability/package.py:61]: Spack has extras, we call them "variants". So if you want to make them optional, you can. Again, I'll leave it up to you, I'm a PyTorch guy, not a TF guy. I don't know much about the TF ecosystem, but will try to help where I can to maintain it.

**Review by adamjstewart — DISMISSED** [review]
I'm fine with the PR as is, but will await your reply before merging in case you want to change anything.

File: var/spack/repos/builtin/packages/py-tensorflow-probability/package.py
Comment (jonas-eschle) [reply var/spack/repos/builtin/packages/py-tensorflow-probability/package.py]: That makes a lot of sense, with that policy, I fully agree, marked for deprecation.

File: var/spack/repos/builtin/packages/py-tensorflow-probability/package.py
Comment (jonas-eschle) [reply var/spack/repos/builtin/packages/py-tensorflow-probability/package.py:61]: Ah, I get it know, thanks, I was looking for something else in the docs (more like "optional deps"). Reconsidering, as not even TFP has any as a dep, let's make them extras too! Can you quickly check if this is correct, i.e. I am not fully sure if now jax _and_ tf can be given as an extra at the same time. They can, right?

For the version, same story: if people can restrict it, then let them do that (unless we're aware of specific incompatibilities, no need to be ignorant). I've relaxed it, restricted it to the major version and leave the rest floating, TFP also does only a runtime check ATM.

File: var/spack/repos/builtin/packages/py-tensorflow-probability/package.py
Comment (jonas-eschle) [reply var/spack/repos/builtin/packages/py-tensorflow-probability/package.py]: As said in the other, thread, agree, especially if it's easy to set limits. We can still impose limits if we're aware of incompatibilities, but so far, people using pip face the same issue -> relaxed

**Review by jonas-eschle — COMMENTED** [review]
thanks for all the introduction, things make a lot more sense to me now!

File: var/spack/repos/builtin/packages/py-tensorflow-probability/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-tensorflow-probability/package.py]: TF 1.14 requires Python 3.7 and older, which Spack no longer supports. If it's not possible to build this version with newer TF, we can just remove this version.

Comment (jonas-eschle) [reply var/spack/repos/builtin/packages/py-tensorflow-probability/package.py]: Removed, and deprecated the 2.4 (that is ~4 years old)

**Comment by jonas-eschle** [comment]
Ready from my side

**Comment by adamjstewart** [comment]
No idea what's going on with the tests. Trying to rerun...

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
FYI, every time you update this branch, it has to wait for GitLab CI to finish on develop before it can even start running on this PR. If you're just trying to re-run CI, it's better to close and reopen the PR.

**Pull request merged**
